File: tests/fixtures/typescript/auth_service.ts
// Target: bookmark an interface
export interface Claims {
  subject: string;
  expiry: Date;
  roles: string[];
}

// Target: bookmark an enum
export enum AuthError {
  InvalidToken = "INVALID_TOKEN",
  Expired = "EXPIRED",
  InsufficientPermissions = "INSUFFICIENT_PERMISSIONS",
}

// Target: bookmark a class declaration
export class AuthService {
  private secret: string;
  private issuer: string;
  private tokenCache: Map<string, Claims> = new Map();

  constructor(secret: string, issuer: string = "codemark") {
    this.secret = secret;
    this.issuer = issuer;
  }

  // Target: bookmark a method
  async validateToken(token: string): Promise<Claims> {
    const cached = this.tokenCache.get(token);
    if (cached) {
      if (cached.expiry <= new Date()) {
        this.tokenCache.delete(token);
        throw new Error(AuthError.Expired);
      }
      return cached;
    }

    const claims = this.decode(token);
    if (claims.expiry <= new Date()) {
      throw new Error(AuthError.Expired);
    }

    this.tokenCache.set(token, claims);
    return claims;
  }

  // Target: bookmark a private method
  private decode(token: string): Claims {
    if (!token) {
      throw new Error(AuthError.InvalidToken);
    }
    return {
      subject: "user-1",
      expiry: new Date(Date.now() + 3600000),
      roles: ["admin"],
    };
  }

  // Target: bookmark a method with complex params
  checkPermission(claims: Claims, required: string, resource: string): void {
    if (!claims.roles.includes(required)) {
      throw new Error(`${AuthError.InsufficientPermissions}: ${required} on ${resource}`);
    }
  }
}

// Target: bookmark an arrow function
export const createDefaultAuthService = (): AuthService => {
  return new AuthService("default-secret");
};

// Target: bookmark a regular function
export function validateAndCheck(
  service: AuthService,
  token: string,
  requiredRole: string
): Promise<Claims> {
  return service.validateToken(token).then((claims) => {
    service.checkPermission(claims, requiredRole, "*");
    return claims;
  });
}

// Target: bookmark a type alias
export type AuthProvider = {
  validateToken(token: string): Promise<Claims>;
  refreshToken(token: string): Promise<string>;
};

File: tests/fixtures/rust/auth_service.rs
use std::collections::HashMap;
use std::time::{Duration, SystemTime};

// Target: bookmark a struct declaration
#[derive(Debug, Clone)]
pub struct Claims {
    pub subject: String,
    pub expiry: SystemTime,
    pub roles: Vec<String>,
}

// Target: bookmark an enum with variants
#[derive(Debug, thiserror::Error)]
pub enum AuthError {
    #[error("invalid token")]
    InvalidToken,
    #[error("token expired")]
    Expired,
    #[error("insufficient permissions: requires {required}")]
    InsufficientPermissions { required: String },
}

// Target: bookmark a trait declaration
pub trait AuthProvider {
    fn validate_token(&self, token: &str) -> Result<Claims, AuthError>;
    fn refresh_token(&self, token: &str) -> Result<String, AuthError>;
}

// Target: bookmark a struct with fields
pub struct AuthService {
    secret: String,
    issuer: String,
    token_cache: HashMap<String, Claims>,
}

// Target: bookmark the impl block constructor
impl AuthService {
    pub fn new(secret: String, issuer: String) -> Self {
        AuthService {
            secret,
            issuer,
            token_cache: HashMap::new(),
        }
    }

    // Target: bookmark a private method inside impl
    fn decode(&self, token: &str) -> Result<Claims, AuthError> {
        if token.is_empty() {
            return Err(AuthError::InvalidToken);
        }
        // Simulated decode
        Ok(Claims {
            subject: "user-1".to_string(),
            expiry: SystemTime::now() + Duration::from_secs(3600),
            roles: vec!["admin".to_string()],
        })
    }

    fn encode(&self, claims: &Claims) -> Result<String, AuthError> {
        Ok(format!("{}:{}:{}", self.issuer, claims.subject, self.secret))
    }

    // Target: bookmark a method with complex parameters
    pub fn check_permission(
        &self,
        claims: &Claims,
        required: &str,
        resource: &str,
    ) -> Result<(), AuthError> {
        if !claims.roles.contains(&required.to_string()) {
            return Err(AuthError::InsufficientPermissions {
                required: format!("{required} on {resource}"),
            });
        }
        Ok(())
    }
}

// Target: bookmark a trait implementation
impl AuthProvider for AuthService {
    fn validate_token(&self, token: &str) -> Result<Claims, AuthError> {
        if let Some(cached) = self.token_cache.get(token) {
            if cached.expiry > SystemTime::now() {
                return Ok(cached.clone());
            }
        }
        let claims = self.decode(token)?;
        if claims.expiry <= SystemTime::now() {
            return Err(AuthError::Expired);
        }
        Ok(claims)
    }

    fn refresh_token(&self, token: &str) -> Result<String, AuthError> {
        let claims = self.decode(token)?;
        if claims.expiry
            <= SystemTime::now()
                .checked_sub(Duration::from_secs(3600))
                .unwrap_or(SystemTime::UNIX_EPOCH)
        {
            return Err(AuthError::Expired);
        }
        self.encode(&claims)
    }
}

// Target: bookmark a free function
pub fn create_default_auth_service() -> AuthService {
    AuthService::new("default-secret".to_string(), "codemark".to_string())
}

// Target: bookmark a generic function
pub fn validate_and_check<P: AuthProvider>(
    provider: &P,
    token: &str,
    required_role: &str,
) -> Result<Claims, AuthError> {
    let claims = provider.validate_token(token)?;
    if !claims.roles.contains(&required_role.to_string()) {
        return Err(AuthError::InsufficientPermissions {
            required: required_role.to_string(),
        });
    }
    Ok(claims)
}
